File: cpp/07 - interest-is-interesting/interest_is_interesting.cpp
// interest_rate returns the interest rate for the provided balance.
double interest_rate(double balance) {
    if (balance < 0) {
        return 3.213;
    }
    else if (balance >= 0 && balance < 1000) {
        return 0.5;
    }
    else if (balance >= 1000 && balance < 5000) {
        return 1.621;
    }
    else if (balance >= 5000) {
        return 2.475;
    }
    else {
        return balance;
    }
}

// yearly_interest calculates the yearly interest for the provided balance.
double yearly_interest(double balance) {
    return balance * (interest_rate(balance)/100);
}

// annual_balance_update calculates the annual balance update, taking into
// account the interest rate.
double annual_balance_update(double balance) {
    // TODO: Implement the annual_balance_update function
    return balance + yearly_interest(balance);
}

// years_until_desired_balance calculates the minimum number of years required
// to reach the desired balance.
int years_until_desired_balance(double balance, double target_balance) {
    int years = 0;
    while (balance < target_balance) {
        balance = annual_balance_update(balance);
        years += 1;
    }
    return years;
}

File: csharp/06 - interest-is-interesting/InterestIsInteresting.cs
using System;

static class SavingsAccount
{
    public static float InterestRate(decimal balance)
    {
        if (balance > 5000)
        {
            return 2.475f;
        }
        else if (balance > 1000)
        {
            return 1.621f;
        }
        else if (balance < 1000 && balance > 0)
        {
            return 0.5f;
        }
        else if (balance < 0)
        {
            return 3.213f;
        }
    }

    public static decimal Interest(decimal balance) => (balance * (decimal)InterestRate(balance)/100);

    public static decimal AnnualBalanceUpdate(decimal balance) => balance += Interest(balance);

    public static int YearsBeforeDesiredBalance(decimal balance, decimal targetBalance)
    {
        int years = 0;
        while (balance < targetBalance)
        {
            balance = (decimal)AnnualBalanceUpdate(balance);
            years++;
        }
        return years;
    }
}
